core/src/main/java/com/linecorp/armeria/client/Http1ResponseDecoder.java

```diff
@@ -263,4 +279,47 @@ public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exceptio
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
         ctx.fireExceptionCaught(cause);
     }
+
+    void setKeepAliveHandler(ChannelHandlerContext ctx, KeepAliveHandler keepAliveHandler) {
+        this.keepAliveHandler = keepAliveHandler;
+        maybeInitializeKeepAliveHandler(ctx);
+    }
+
+    private void maybeInitializeKeepAliveHandler(ChannelHandlerContext ctx) {
+        if (keepAliveHandler != null && ctx.channel().isActive()) {
+            keepAliveHandler.initialize(ctx);
+        }
+    }
+
+    private void destroyKeepAliveHandler() {
+        if (keepAliveHandler != null) {
+            keepAliveHandler.destroy();
+        }
+    }
+
+    private void onPingRead(Object msg) {
+        if (msg instanceof LastHttpContent) {
+            onPingComplete();
+        } else {
+            assert keepAliveHandler != null;
+            keepAliveHandler.onPing();
+        }
```

Comment:
I think it's not very common but:

- An `OPTIONS` response can have chunked content.
- HTTP/1 decoder may emit a `FullHttpResponse` which is both an `HttpResponse` and an `LastHttpContent`.

In such cases, `keepAliveHandler.onPing()` can be invoked more or less than desired.

Reply:
Good point, I didn't think of that.